File: ts-api/examples/example_augment.ts
/**
 * NovelAI Augment & Upscale API 使用例
 * 
 * 画像加工ツール（カラー化、表情変換、スケッチ化など）とアップスケール機能のサンプルコード
 * 
 * 注意: width/height は画像から自動検出されるため、指定不要です
 */
import { NovelAIClient } from '../src/client';
import path from 'path';
import dotenv from 'dotenv';

// .envファイルを読み込み
dotenv.config();

const OUTPUT_DIR = path.join(__dirname, '..', 'output', 'augment');

async function main() {
  const client = new NovelAIClient();

  // アンラス残高を確認
  const balance = await client.getAnlasBalance();
  console.log(`\n📊 現在のアンラス残高: ${balance.total}`);
  console.log(`   (固定: ${balance.fixed}, 購入済み: ${balance.purchased})\n`);

  // =====================================================
  // 1. カラー化 (colorize)
  // =====================================================
  console.log('🎨 カラー化テスト...');
  try {
    const colorizeResult = await client.augmentImage({
      req_type: "colorize",
      image: path.join(__dirname, '..', 'reference', 'input.jpeg'),  // モノクロ画像
      // width/height は自動検出
      prompt: "vibrant colors, detailed shading",  // カラー化のヒント
      defry: 3,  // 中程度の変更 (0=最強, 5=最弱)
      save_dir: OUTPUT_DIR,
    });
    console.log(`   ✅ 保存先: ${colorizeResult.saved_path}`);
    console.log(`   💰 消費アンラス: ${colorizeResult.anlas_consumed ?? 'N/A'}`);
  } catch (error) {
    console.log(`   ❌ エラー: ${error}`);
  }

  // =====================================================
  // 2. 表情変換 (emotion)
  // =====================================================
  console.log('\n😊 表情変換テスト...');

  // 利用可能な表情キーワード:
  // neutral, happy, sad, angry, scared, surprised, tired, excited,
  // nervous, thinking, confused, shy, disgusted, smug, bored,
  // laughing, irritated, aroused, embarrassed, love, worried,
  // determined, hurt, playful

  try {
    const emotionResult = await client.augmentImage({
      req_type: "emotion",
      image: path.join(__dirname, '..', 'reference', 'input.jpeg'),  // 顔画像
      // width/height は自動検出
      prompt: "happy",  // 表情キーワード
      defry: 0,  // 最強の変更
      save_dir: OUTPUT_DIR,
    });
    console.log(`   ✅ 保存先: ${emotionResult.saved_path}`);
    console.log(`   💰 消費アンラス: ${emotionResult.anlas_consumed ?? 'N/A'}`);
  } catch (error) {
    console.log(`   ❌ エラー: ${error}`);
  }

  // =====================================================
  // 3. スケッチ化 (sketch)
  // =====================================================
  console.log('\n✏️ スケッチ化テスト...');
  try {
    const sketchResult = await client.augmentImage({
      req_type: "sketch",
      image: path.join(__dirname, '..', 'reference', 'input.jpeg'),
      // width/height は自動検出
      save_dir: OUTPUT_DIR,
    });
    console.log(`   ✅ 保存先: ${sketchResult.saved_path}`);
    console.log(`   💰 消費アンラス: ${sketchResult.anlas_consumed ?? 'N/A'}`);
  } catch (error) {
    console.log(`   ❌ エラー: ${error}`);
  }

  // =====================================================
  // 4. 線画抽出 (lineart)
  // =====================================================
  console.log('\n📝 線画抽出テスト...');
  try {
    const lineartResult = await client.augmentImage({
      req_type: "lineart",
      image: path.join(__dirname, '..', 'reference', 'input.jpeg'),
      // width/height は自動検出
      save_dir: OUTPUT_DIR,
    });
    console.log(`   ✅ 保存先: ${lineartResult.saved_path}`);
    console.log(`   💰 消費アンラス: ${lineartResult.anlas_consumed ?? 'N/A'}`);
  } catch (error) {
    console.log(`   ❌ エラー: ${error}`);
  }

  // =====================================================
  // 5. デクラッター (declutter)
  // =====================================================
  console.log('\n🧹 デクラッターテスト...');
  try {
    const declutterResult = await client.augmentImage({
      req_type: "declutter",
      image: path.join(__dirname, '..', 'reference', 'input.jpeg'),
      // width/height は自動検出
      save_dir: OUTPUT_DIR,
    });
    console.log(`   ✅ 保存先: ${declutterResult.saved_path}`);
    console.log(`   💰 消費アンラス: ${declutterResult.anlas_consumed ?? 'N/A'}`);
  } catch (error) {
    console.log(`   ❌ エラー: ${error}`);
  }

  // =====================================================
  // 6. 背景除去 (bg-removal) - 常にアンラス消費
  // =====================================================
  console.log('\n🖼️ 背景除去テスト（⚠️ 常にアンラス消費）...');
  try {
    const bgRemovalResult = await client.augmentImage({
      req_type: "bg-removal",
      image: path.join(__dirname, '..', 'reference', 'input.jpeg'),
      // width/height は自動検出
      save_dir: OUTPUT_DIR,
    });
    console.log(`   ✅ 保存先: ${bgRemovalResult.saved_path}`);
    console.log(`   💰 消費アンラス: ${bgRemovalResult.anlas_consumed ?? 'N/A'}`);
  } catch (error) {
    console.log(`   ❌ エラー: ${error}`);
  }

  // =====================================================
  // 7. アップスケール (upscale) - 常にアンラス消費
  // =====================================================
  console.log('\n🔍 アップスケールテスト（⚠️ 常にアンラス消費）...');
  try {
    const upscaleResult = await client.upscaleImage({
      image: path.join(__dirname, '..', 'reference', 'input.jpeg'),
      // width/height は自動検出
      scale: 4,  // 4倍拡大 (2 or 4)
      save_dir: OUTPUT_DIR,
    });
    console.log(`   ✅ 保存先: ${upscaleResult.saved_path}`);
    console.log(`   📐 出力サイズ: ${upscaleResult.output_width}x${upscaleResult.output_height}`);
    console.log(`   💰 消費アンラス: ${upscaleResult.anlas_consumed ?? 'N/A'}`);
  } catch (error) {
    console.log(`   ❌ エラー: ${error}`);
  }

  // 最終アンラス残高を確認
  const finalBalance = await client.getAnlasBalance();
  console.log(`\n📊 最終アンラス残高: ${finalBalance.total}`);
  console.log(`   総消費: ${balance.total - finalBalance.total}\n`);
}

main().catch(console.error);

File: rust-api/examples/example_augment.rs
//! NovelAI Augment & Upscale API Example
//!
//! 画像加工ツール（カラー化、表情変換、スケッチ化など）とアップスケール機能のサンプル
//!
//! 注意: width/height は画像から自動検出されるため、指定不要です
//!
//! Run with: cargo run --example example_augment

use anyhow::Result;
use novelai_api::client::NovelAIClient;
use novelai_api::constants::AugmentReqType;
use novelai_api::schemas::*;

#[tokio::main]
async fn main() -> Result<()> {
    dotenvy::dotenv().ok();

    let output_dir = "output/augment";
    std::fs::create_dir_all(output_dir)?;

    let client = NovelAIClient::new(None, None)?;
    let input_image = "reference/input.jpeg";

    // アンラス残高を確認
    let balance = client.get_anlas_balance().await?;
    println!("\n現在のアンラス残高: {}", balance.total);
    println!("  (固定: {}, 購入済み: {})\n", balance.fixed, balance.purchased);

    // =====================================================
    // 1. カラー化 (colorize)
    // =====================================================
    println!("カラー化テスト...");
    match client
        .augment_image(&AugmentParams {
            req_type: AugmentReqType::Colorize,
            image: ImageInput::FilePath(input_image.into()),
            prompt: Some("vibrant colors, detailed shading".into()),
            defry: Some(3),
            save: SaveTarget::Directory { dir: output_dir.into(), filename: None },
        })
        .await
    {
        Ok(result) => {
            println!(
                "  保存先: {}",
                result.saved_path.as_deref().unwrap_or("N/A")
            );
            println!(
                "  消費アンラス: {}",
                result
                    .anlas_consumed
                    .map_or("N/A".into(), |v| v.to_string())
            );
        }
        Err(e) => println!("  エラー: {}", e),
    }

    // =====================================================
    // 2. 表情変換 (emotion)
    // =====================================================
    println!("\n表情変換テスト...");
    match client
        .augment_image(&AugmentParams {
            req_type: AugmentReqType::Emotion,
            image: ImageInput::FilePath(input_image.into()),
            prompt: Some("happy".into()),
            defry: Some(0),
            save: SaveTarget::Directory { dir: output_dir.into(), filename: None },
        })
        .await
    {
        Ok(result) => {
            println!(
                "  保存先: {}",
                result.saved_path.as_deref().unwrap_or("N/A")
            );
            println!(
                "  消費アンラス: {}",
                result
                    .anlas_consumed
                    .map_or("N/A".into(), |v| v.to_string())
            );
        }
        Err(e) => println!("  エラー: {}", e),
    }

    // =====================================================
    // 3. スケッチ化 (sketch)
    // =====================================================
    println!("\nスケッチ化テスト...");
    match client
        .augment_image(&AugmentParams {
            req_type: AugmentReqType::Sketch,
            image: ImageInput::FilePath(input_image.into()),
            prompt: None,
            defry: None,
            save: SaveTarget::Directory { dir: output_dir.into(), filename: None },
        })
        .await
    {
        Ok(result) => {
            println!(
                "  保存先: {}",
                result.saved_path.as_deref().unwrap_or("N/A")
            );
            println!(
                "  消費アンラス: {}",
                result
                    .anlas_consumed
                    .map_or("N/A".into(), |v| v.to_string())
            );
        }
        Err(e) => println!("  エラー: {}", e),
    }

    // =====================================================
    // 4. 線画抽出 (lineart)
    // =====================================================
    println!("\n線画抽出テスト...");
    match client
        .augment_image(&AugmentParams {
            req_type: AugmentReqType::Lineart,
            image: ImageInput::FilePath(input_image.into()),
            prompt: None,
            defry: None,
            save: SaveTarget::Directory { dir: output_dir.into(), filename: None },
        })
        .await
    {
        Ok(result) => {
            println!(
                "  保存先: {}",
                result.saved_path.as_deref().unwrap_or("N/A")
            );
            println!(
                "  消費アンラス: {}",
                result
                    .anlas_consumed
                    .map_or("N/A".into(), |v| v.to_string())
            );
        }
        Err(e) => println!("  エラー: {}", e),
    }

    // =====================================================
    // 5. デクラッター (declutter)
    // =====================================================
    println!("\nデクラッターテスト...");
    match client
        .augment_image(&AugmentParams {
            req_type: AugmentReqType::Declutter,
            image: ImageInput::FilePath(input_image.into()),
            prompt: None,
            defry: None,
            save: SaveTarget::Directory { dir: output_dir.into(), filename: None },
        })
        .await
    {
        Ok(result) => {
            println!(
                "  保存先: {}",
                result.saved_path.as_deref().unwrap_or("N/A")
            );
            println!(
                "  消費アンラス: {}",
                result
                    .anlas_consumed
                    .map_or("N/A".into(), |v| v.to_string())
            );
        }
        Err(e) => println!("  エラー: {}", e),
    }

    // =====================================================
    // 6. 背景除去 (bg-removal) - 常にアンラス消費
    // =====================================================
    println!("\n背景除去テスト（常にアンラス消費）...");
    match client
        .augment_image(&AugmentParams {
            req_type: AugmentReqType::BgRemoval,
            image: ImageInput::FilePath(input_image.into()),
            prompt: None,
            defry: None,
            save: SaveTarget::Directory { dir: output_dir.into(), filename: None },
        })
        .await
    {
        Ok(result) => {
            println!(
                "  保存先: {}",
                result.saved_path.as_deref().unwrap_or("N/A")
            );
            println!(
                "  消費アンラス: {}",
                result
                    .anlas_consumed
                    .map_or("N/A".into(), |v| v.to_string())
            );
        }
        Err(e) => println!("  エラー: {}", e),
    }

    // =====================================================
    // 7. アップスケール (upscale) - 常にアンラス消費
    // =====================================================
    println!("\nアップスケールテスト（常にアンラス消費）...");
    match client
        .upscale_image(&UpscaleParams {
            image: ImageInput::FilePath(input_image.into()),
            scale: 4,
            save: SaveTarget::Directory { dir: output_dir.into(), filename: None },
        })
        .await
    {
        Ok(result) => {
            println!(
                "  保存先: {}",
                result.saved_path.as_deref().unwrap_or("N/A")
            );
            println!(
                "  出力サイズ: {}x{}",
                result.output_width, result.output_height
            );
            println!(
                "  消費アンラス: {}",
                result
                    .anlas_consumed
                    .map_or("N/A".into(), |v| v.to_string())
            );
        }
        Err(e) => println!("  エラー: {}", e),
    }

    // 最終アンラス残高を確認
    let final_balance = client.get_anlas_balance().await?;
    println!("\n最終アンラス残高: {}", final_balance.total);
    println!(
        "  総消費: {}\n",
        balance.total.saturating_sub(final_balance.total)
    );

    Ok(())
}
